Minor fixes (#10)
eficode-academy/rf-advanced-web-katas: exercises/03-xpath-intro.md

Ideally, web development and test automation goes hand in hand, so
that the people who develop the websites, would also do (at least
a bit of) test automation, or be aware of it. However, more often than not this is not
the case and that causes trouble for test automation developers.
Websites are often created with "UX first" approach, which is good for the users,

> but it typically means more work for the test automation developers.

Reviewer: I feel this is controversial statement. In my view, messy UI code is precisely caused by *not* creating the application with "UX first" approach. 

Usually when things are designed and thought through before hand, the frontend code is clean with good element id's & classes and thus, easily testable. It is precisely people hacking together some crap that results in spaghetti javascript, invalid HTML, weird `<iframe>`'s and the whole shebang. 

This is also the reason why it would be good idea for the developers to write test automation themselves -- because then they would have to "eat their own dog food" ie. deal with the spaghetti they have created.

Reviewer 2: That is a good point. The paragraph could be rephrased to something like

> Unfortunately, websites are often created with little emphasis on automation, which means features work perfectly fine for a human user, but are very tedious to automate.

Nowadays, websites are created with modern frameworks, which generate
nice looking websites from templates. Pretty often developers
forget to add unique `id` attributes to elements, so we need
alternative locators for our elements.

> This is where XPaths step in. Sure enough, there are other locator strategies

Reviewer: I would very much drive here the point home that you should pretty much always use some other locator strategies than XPath. It is only needed when the only identifiable information of an element is the visible text in the element. 

(With Browser, this gets even easier with composable selectors, so you virtually never need XPath)

Identifying element in any other way usually means you avoid XPath. I agree that the training should have this section, but I would just emphasise here that XPath should be last resort.

> with SeleniumLibrary and Browser than XPaths and IDs (like CSS paths), but quite often using an XPath is the
> only way to find both a functional and unique identifier for an element.

Reviewer: Hard disagree with:
> but quite often 

Rather:
```suggestion
with SeleniumLibrary and Browser than XPaths and IDs (like CSS paths), but when using an XPath is the
only way to find both a functional and unique identifier for an element.
```

Author: Ok.

The trick to XPaths is to make a locator which is accurate enough to pinpoint

> your element, but not too exact as to make it fragile, and fast to break.

Reviewer: ```suggestion
your element, but not too exact as to make it fragile, and thus fast to break unnecessarily.
```

Author: Ok.

If a website is likely to change, using bad XPath strategies is an easy way to find
ourself in a swamp of refactoring and debugging. A good mindset for XPaths
is to keep them short, but understandable. Shorter XPaths mean fewer places
for it to break and less time wasted on unnecessary debugging. If they do break,
as long as they still make sense, you can make replacement locators.

A few general tips for good XPath usage:

1. Don't use the browser's `Copy XPath` button. This will give you too
obscure, segmented and precise XPath. For example, Firefox's XPath for the
Bad Flask App's dropdown menu is `/html/body/section[1]/nav/div/a`, which is long,
has 6 strictly ordered parts of generic type, and uses no unique attributes. Any
changes to element types, the page's structure, or the order of those elements
will break the locator. Also, since all elements should be behind `/html/body`, good

> XPaths should **always** simply start with `//` when testing websites.

Reviewer: I feel we could justify why here:
```suggestion
XPaths should **always** simply start with `//` when testing websites, as the operator searches from anywhere in the HTML document.
```

Author: Ok.

will open in our test case or not. While we're looking at the Bad Flask App, let's
open our developer console by right-clicking anywhere on the screen and selecting `inspect`.
It's a good idea to keep the developer console opened always when you're writing Selenium tests.
We notice, that the page's elements don't have `id` fields that would allow us to
easily access them.

Let's start by observing the page's elements and their attributes. We notice that the dropdown button is an
`a` element, which has classes we could use, for example `dropdown-toggle`. However, there's a hidden
element before in the HTML, so we can't use that `a` directly. Instead, we can use its parent
`div` element to handle the click. Also, it has a class called `open` when the dropdown is opened and it's
missing when it's closed. So, in other words _if_ the `div` element has a class called `open`, we should click it.

> Still, the locator we got may sound a bit cryptic. Instead of adding it directly into our keyword, let's

Reviewer: ```suggestion
Still, the locator we got may look a bit cryptic. Instead of adding it directly into our keyword, let's
```
